File: src/browser_driver.mjs
// Embedded by the Horde binary. The app provides Playwright as a pinned dev dependency.
import readline from 'node:readline';
import { createRequire } from 'node:module';
import path from 'node:path';

const input = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
const iterator = input[Symbol.asyncIterator]();
const send = value => process.stdout.write(`${JSON.stringify(value)}\n`);
const next = async () => {
  const row = await iterator.next();
  if (row.done) throw new Error('browser controller closed');
  if (row.value.length > 64 * 1024) throw new Error('browser command too large');
  return JSON.parse(row.value);
};
let browser, context, page, origin, screenshot;
let version = 0;
let current = new Map();
let external = [];
const clip = (value, size = 256) => String(value ?? '').replace(/\s+/g, ' ').trim().slice(0, size);
async function screenshotIfPossible() {
  if (page && screenshot) {
    try { await page.screenshot({ path: screenshot, fullPage: false, timeout: 3000 }); } catch { /* evidence is best effort */ }
  }
}
async function observe(ack = null) {
  if (new URL(page.url()).origin !== origin) throw new Error('browser left the app origin');
  const locator = page.locator('button, a[href], input, select, textarea, [role="button"], [role="link"]');
  const count = await locator.count();
  if (count > 128) throw new Error('page has more than 128 controls');
  const controls = [];
  const map = new Map();
  for (let n = 0; n < count; n += 1) {
    const handle = await locator.nth(n).elementHandle();
    if (!handle || !(await handle.isVisible()) || !(await handle.isEnabled())) continue;
    const info = await handle.evaluate(element => {
      const tag = element.tagName.toLowerCase();
      const type = (element.getAttribute('type') ?? '').toLowerCase();
      const role = (element.getAttribute('role') ?? '').toLowerCase();
      const label = element.getAttribute('aria-label') || element.labels?.[0]?.innerText || element.getAttribute('placeholder') || element.innerText || element.getAttribute('value') || '';
      const operations = tag === 'input' && type === 'file' ? [] :
        (tag === 'select' ? ['select'] : ((tag === 'textarea' || (tag === 'input' && !['button', 'submit', 'checkbox', 'radio', 'hidden', 'file'].includes(type))) ? ['fill'] : ['click']));
      return { label, operations, fingerprint: [tag, type, role, element.getAttribute('href'), element.getAttribute('name'), label].join('|') };
    });
    if (info.operations.length === 0) { external.push('file upload control unsupported'); continue; }
    const id = `c${n}`;
    controls.push({ id, label: clip(info.label), operations: info.operations });
    map.set(id, { handle, fingerprint: info.fingerprint, operations: info.operations });
  }
  version += 1;
  current = map;
  send({ kind: 'observation', ack, version, url: page.url(), controls, unsupported: external.splice(0) });
}
async function act(command) {
  const target = current.get(command.id);
  if (command.version !== version || !target || !target.operations.includes(command.operation)) throw new Error('stale or unsupported browser target');
  const fresh = await target.handle.evaluate(element => {
    if (!element.isConnected) return null;
    const tag = element.tagName.toLowerCase();
    const type = (element.getAttribute('type') ?? '').toLowerCase();
    const role = (element.getAttribute('role') ?? '').toLowerCase();
    const label = element.getAttribute('aria-label') || element.labels?.[0]?.innerText || element.getAttribute('placeholder') || element.innerText || element.getAttribute('value') || '';
    return [tag, type, role, element.getAttribute('href'), element.getAttribute('name'), label].join('|');
  });
  if (fresh !== target.fingerprint || !(await target.handle.isVisible()) || !(await target.handle.isEnabled())) throw new Error('browser target changed since observation');
  if (command.operation === 'click') await target.handle.click({ timeout: 5000 });
  else if (command.operation === 'fill' && typeof command.value === 'string') await target.handle.fill(command.value, { timeout: 5000 });
  else if (command.operation === 'select' && typeof command.value === 'string') await target.handle.selectOption(command.value, { timeout: 5000 });
  else throw new Error('browser input requires a declared value');
  await page.waitForTimeout(150);
  await observe(command.decision);
}
async function assertIndependent(assertions, decision) {
  const reports = [];
  for (const item of assertions) {
    let passed = false;
    if (new URL(page.url()).origin !== origin) passed = false;
    else if (item.kind === 'url_path') passed = new URL(page.url()).pathname === item.path;
    else if (item.kind === 'text_visible') passed = await page.getByText(item.text, { exact: true }).first().isVisible().catch(() => false);
    reports.push({ kind: item.kind, passed });
  }
  await screenshotIfPossible();
  send({ kind: 'result', ack: decision, passed: reports.length > 0 && reports.every(row => row.passed) && external.length === 0, assertions: reports });
}
try {
  const start = await next();
  if (start.cmd !== 'start') throw new Error('expected browser start command');
  const requireFromWorkspace = createRequire(path.join(process.cwd(), 'package.json'));
  let chromium;
  try { ({ chromium } = requireFromWorkspace('playwright')); }
  catch { throw new Error('install pinned Playwright in the app workspace and run npx playwright install chromium'); }
  origin = new URL(start.url).origin;
  screenshot = start.screenshot;
  browser = await chromium.launch({ headless: true, env: { PATH: process.env.PATH ?? '', HOME: process.env.HOME ?? '', PLAYWRIGHT_BROWSERS_PATH: process.env.PLAYWRIGHT_BROWSERS_PATH ?? '' } });
  context = await browser.newContext({ serviceWorkers: 'block', permissions: [], acceptDownloads: false });
  await context.route('**/*', route => {
    try { if (new URL(route.request().url()).origin === origin) return route.continue(); }
    catch { /* reject malformed resource URL */ }
    external.push('cross-origin resource or navigation blocked');
    return route.abort();
  });
  if (typeof context.routeWebSocket !== 'function') {
    throw new Error('Playwright 1.48 or newer is required to block cross-origin WebSockets');
  }
  await context.routeWebSocket('**/*', webSocket => {
    try {
      const destination = new URL(webSocket.url());
      if (destination.protocol === 'ws:') destination.protocol = 'http:';
      else if (destination.protocol === 'wss:') destination.protocol = 'https:';
      else throw new Error('unsupported WebSocket scheme');
      if (destination.origin === origin) return webSocket.connectToServer();
    } catch { /* reject malformed WebSocket URL */ }
    external.push('cross-origin WebSocket blocked');
    return webSocket.close();
  });
  page = await context.newPage();
  page.on('popup', popup => { external.push('popup unsupported'); void popup.close(); });
  await page.goto(start.url, { waitUntil: 'domcontentloaded', timeout: 15000 });
  await observe();
  for (;;) {
    const command = await next();
    if (command.cmd === 'act') await act(command);
    else if (command.cmd === 'assert') { await assertIndependent(start.assertions, command.decision); break; }
    else throw new Error('unsupported browser command');
  }
} catch (error) {
  await screenshotIfPossible();
  send({ kind: 'unsupported', reason: clip(error.message, 512) });
} finally {
  if (context) await context.close().catch(() => {});
  if (browser) await browser.close().catch(() => {});
}

File: tests/browser_driver.rs
use serde_json::{Value, json};
use std::{
    io::{BufRead, Write},
    process::{Command, Stdio},
};

const FAKE_PLAYWRIGHT: &str = r#"
const fs = require('node:fs');
let saved = false;
let webSocketHandler;
const element = { tagName: 'BUTTON', labels: [], innerText: 'Save', isConnected: true,
  getAttribute() { return null; } };
const handle = { async isVisible() { return true; }, async isEnabled() { return true; },
  async evaluate(fn) { return fn(element); }, async click() { saved = true; },
  async fill() { throw new Error('unexpected fill'); }, async selectOption() { throw new Error('unexpected select'); } };
const page = { url() { return 'http://127.0.0.1:1234/'; }, async goto() {
    if (process.env.HORDE_TEST_WEBSOCKET) {
      if (!webSocketHandler) throw new Error('WebSocket routing not installed');
      const connections = [];
      for (const url of ['ws://127.0.0.1:1234/ws', 'wss://elsewhere.example/ws']) {
        const route = { url() { return url; }, async close() { connections.push('closed'); },
          connectToServer() { connections.push('connected'); } };
        await webSocketHandler(route);
      }
      if (connections.join(',') !== 'connected,closed') throw new Error('WebSocket origin rule failed');
    }
  }, on() {},
  locator() { return { async count() { return 1; }, nth() { return { async elementHandle() { return handle; } }; } }; },
  async waitForTimeout() {}, getByText() { return { first() { return { async isVisible() { return saved; } }; } }; },
  async screenshot({path}) { fs.writeFileSync(path, Buffer.from('89504e470d0a1a0a','hex')); } };
module.exports = { chromium: { async launch() {
  if (process.env.TYPESAFE_API_KEY) throw new Error('provider key reached driver');
  return { async newContext() { return { async route() {}, async routeWebSocket(_pattern, handler) { webSocketHandler = handler; }, async newPage() { return page; }, async close() {} }; }, async close() {} };
} } };
"#;

fn run_case(first: Value, expect_pass: bool) {
    if Command::new("node").arg("--version").output().is_err() {
        return;
    }
    let temp = tempfile::tempdir().unwrap();
    let workspace = temp.path();
    std::fs::create_dir_all(workspace.join("node_modules/playwright")).unwrap();
    std::fs::write(workspace.join("package.json"), "{}").unwrap();
    std::fs::write(
        workspace.join("node_modules/playwright/index.js"),
        FAKE_PLAYWRIGHT,
    )
    .unwrap();
    std::fs::write(
        workspace.join("driver.mjs"),
        include_str!("../src/browser_driver.mjs"),
    )
    .unwrap();
    let screenshot = workspace.join("shot.png");
    let mut child = Command::new("node")
        .arg(workspace.join("driver.mjs"))
        .current_dir(workspace)
        .env_remove("TYPESAFE_API_KEY")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .spawn()
        .unwrap();
    let mut stdin = child.stdin.take().unwrap();
    let mut stdout = std::io::BufReader::new(child.stdout.take().unwrap());
    writeln!(stdin, "{}", json!({"cmd":"start","url":"http://127.0.0.1:1234/","values":{},"assertions":[{"kind":"text_visible","text":"Saved"}],"screenshot":screenshot})).unwrap();
    stdin.flush().unwrap();
    let mut line = String::new();
    stdout.read_line(&mut line).unwrap();
    let observation: Value = serde_json::from_str(&line).unwrap();
    assert_eq!(observation["kind"], "observation");
    assert_eq!(observation["controls"][0]["id"], "c0");
    let first = if first["cmd"] == "act" {
        json!({"cmd":"act","version":observation["version"],"id":"c0","operation":"click"})
    } else {
        first
    };
    writeln!(stdin, "{first}").unwrap();
    stdin.flush().unwrap();
    line.clear();
    stdout.read_line(&mut line).unwrap();
    let result: Value = serde_json::from_str(&line).unwrap();
    if expect_pass {
        assert_eq!(result["kind"], "observation");
        writeln!(stdin, "{}", json!({"cmd":"assert"})).unwrap();
        stdin.flush().unwrap();
        line.clear();
        stdout.read_line(&mut line).unwrap();
        let result: Value = serde_json::from_str(&line).unwrap();
        assert_eq!(result["kind"], "result");
        assert_eq!(result["passed"], true);
        assert!(screenshot.exists());
    } else {
        assert_eq!(result["kind"], "result");
        assert_eq!(result["passed"], false);
    }
    drop(stdin);
    assert!(child.wait().unwrap().success());
}
#[test]
fn embedded_driver_checks_actions_and_independent_done_assertions() {
    run_case(json!({"cmd":"act"}), true);
    run_case(json!({"cmd":"assert"}), false);
}
#[test]
fn embedded_driver_rejects_stale_control_versions() {
    if Command::new("node").arg("--version").output().is_err() {
        return;
    }
    let temp = tempfile::tempdir().unwrap();
    let workspace = temp.path();
    std::fs::create_dir_all(workspace.join("node_modules/playwright")).unwrap();
    std::fs::write(workspace.join("package.json"), "{}").unwrap();
    std::fs::write(
        workspace.join("node_modules/playwright/index.js"),
        FAKE_PLAYWRIGHT,
    )
    .unwrap();
    std::fs::write(
        workspace.join("driver.mjs"),
        include_str!("../src/browser_driver.mjs"),
    )
    .unwrap();
    let mut child = Command::new("node")
        .arg(workspace.join("driver.mjs"))
        .current_dir(workspace)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .spawn()
        .unwrap();
    let mut stdin = child.stdin.take().unwrap();
    let mut stdout = std::io::BufReader::new(child.stdout.take().unwrap());
    writeln!(stdin, "{}", json!({"cmd":"start","url":"http://127.0.0.1:1234/","values":{},"assertions":[{"kind":"text_visible","text":"Saved"}],"screenshot":null})).unwrap();
    let mut line = String::new();
    stdout.read_line(&mut line).unwrap();
    writeln!(
        stdin,
        "{}",
        json!({"cmd":"act","version":0,"id":"c0","operation":"click"})
    )
    .unwrap();
    line.clear();
    stdout.read_line(&mut line).unwrap();
    let result: Value = serde_json::from_str(&line).unwrap();
    assert_eq!(result["kind"], "unsupported");
    assert!(result["reason"].as_str().unwrap().contains("stale"));
    drop(stdin);
    assert!(child.wait().unwrap().success());
}

#[test]
fn embedded_driver_blocks_cross_origin_websockets() {
    if Command::new("node").arg("--version").output().is_err() {
        return;
    }
    let temp = tempfile::tempdir().unwrap();
    let workspace = temp.path();
    std::fs::create_dir_all(workspace.join("node_modules/playwright")).unwrap();
    std::fs::write(workspace.join("package.json"), "{}").unwrap();
    std::fs::write(
        workspace.join("node_modules/playwright/index.js"),
        FAKE_PLAYWRIGHT,
    )
    .unwrap();
    std::fs::write(
        workspace.join("driver.mjs"),
        include_str!("../src/browser_driver.mjs"),
    )
    .unwrap();
    let mut child = Command::new("node")
        .arg(workspace.join("driver.mjs"))
        .current_dir(workspace)
        .env("HORDE_TEST_WEBSOCKET", "1")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .spawn()
        .unwrap();
    let mut stdin = child.stdin.take().unwrap();
    let mut stdout = std::io::BufReader::new(child.stdout.take().unwrap());
    writeln!(stdin, "{}", json!({"cmd":"start","url":"http://127.0.0.1:1234/","assertions":[{"kind":"url_path","path":"/"}],"screenshot":null})).unwrap();
    let mut line = String::new();
    stdout.read_line(&mut line).unwrap();
    let observation: Value = serde_json::from_str(&line).unwrap();
    assert_eq!(observation["kind"], "observation");
    assert_eq!(
        observation["unsupported"][0],
        "cross-origin WebSocket blocked"
    );
    drop(stdin);
    assert!(child.wait().unwrap().success());
}
